### ISSUE TITLE
AdminActionForm with access to the queryset?

### ISSUE THREAD

**Issue by matthiask** [michalpokusa/django-admin-action-forms#2]
It would be useful to add fields or choices to the action form depending on the selected objects. 

I don't have a clear idea how the API might look, maybe the form could get the queryset as an init parameter? Something like 

```python
def __init__(self, *args, **kwargs):
    self.queryset = kwargs.pop("queryset")
    super().__init__(*args, **kwargs)
    # ...
```

Of course that wouldn't be backwards compatible with existing action forms.

For now I'm doing the ugly thing of inspecting the call stack by using `inspect.currentframe().f_back.f_back....` etc. It works very well but it's brittle since it depends on the exact spelling of variables in upper layers. And just to reiterate, it's ugly :-)

Feel free to close this issue if something like this is out of scope.

**Comment by michalpokusa** [comment]
I was thinking about something similar, where the form can be customized by the `request`, although the `queryset` can be derived from it. Can you give an example of how the queryset might influence the action form? I want to understant the use case.

Currently the form is being modified at runtime as field are being removed from it based on `fields`/`get_fields`... and widget for specific date related fields are being replaced, but right now it is in private methods. I can see using the same technique for what you described.

I have an idea how to do it, but I would have to change the autocomplete view and keep the compatibility with `django-no-queryset-admin-actions`.

I will try some things and update this issue with results.

**Comment by matthiask** [comment]
I'm posting the full unredacted `AdminActionForm` I'm working with currently, hopefully it clarifies things!

I have textbooks with chapters; the list view is pre-filtered by textbook, and each textbook has its own list of preferences for chapters. So, the available preferences vary by textbook, but are the same for each chapter in the queryset. Each chapter can have different preference values, but I want to use an admin action to allow assigning values to the chapters in bulk.

The following admin action form creates a form which allows changing the template used by each chapter, and also allows setting preferences on all selected chapters at once:

```python
class BulkUpdateActionForm(AdminActionForm):
    class Meta:
        list_objects = True

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        f = Chapter._meta.get_field("template_key").formfield(
            initial="", required=False
        )

        modeladmin = truthy_variable_from_stack("modeladmin")

        self.fields["template_key"] = f
        f.choices = [
            ("", _("No change")),
            *[
                choice
                for choice in f.choices
                if choice[0].startswith(modeladmin.book_series_identifier)
            ],
        ]

        # Try fetching at least one chapter:
        if queryset := truthy_variable_from_stack("queryset"):
            chapter = queryset[0]

            for preference in chapter.textbook.preferences.all():
                name = f"preference_{preference.key}"
                self.fields[name] = forms.ChoiceField(
                    label=capfirst(preference.key.replace("_", " ")),
                    choices=[
                        ("", _("No change")),
                        *(
                            (choice, capfirst(choice.replace("_", " ")))
                            for choice in preference.choices.splitlines()
                        ),
                    ],
                    initial="",
                    required=False,
                )
```

I'm omitting the action itself, I basically take the queryset and assign the selected template and the preference values on all objects in there.

I'd like to have a better way to find at least one chapter in the queryset so that I could go do the textbook to find the available preferences.

I hope that clarifies things :)

**Comment by michalpokusa** [comment]
@matthiask I just released version 1.2.0 that adds [`ActionForm.__post_init__`](https://github.com/michalpokusa/django-admin-action-forms?tab=readme-ov-file#def-__post_init__modeladmin-request-queryset) method. In my understanding this should solve what you are doing with the `inspect` module.

If you could, please test your use case and let me know about the results. Thanks.
